node_exporter.go

```diff
@@ -121,6 +122,7 @@ func (h *handler) innerHandler(filters ...string) (http.Handler, error) {
 	handler := promhttp.HandlerFor(
 		prometheus.Gatherers{h.exporterMetricsRegistry, r},
 		promhttp.HandlerOpts{
+			ErrorLog:            stdlog.New(log.NewStdlibAdapter(level.Error(h.logger)), "", 0),
```

Comment:
I've tested this, and I found that the resulting log lines are truncated at the first newline, which leads to missing message lines about the errors nested within a MultiError. For example, the 6 errors that are suppose to show after this log line do not show in the log:
```
level=error ts=2020-11-15T00:30:24.336Z caller=stdlib.go:89 msg="error gathering metrics: 6 error(s) occurred:"
```
This log entry originated from [client_golang/prometheus/promhttp/http.go#L129](https://github.com/prometheus/client_golang/blob/06b1a0a6ae29dd8b39953dc7f1954a0b2fd680be/prometheus/promhttp/http.go#L129) and the MultiError is formatted with newlines at [client_golang/prometheus/registry.go#L225](https://github.com/prometheus/client_golang/blob/06b1a0a6ae29dd8b39953dc7f1954a0b2fd680be/prometheus/registry.go#L225). Apparently everything after the first newline is truncated by [logRegexpMsg defined at kit/log/stdlib.go#L99](https://github.com/go-kit/kit/blob/1e9fdf6a9f7e74e18922e858194e481c7d97c52a/log/stdlib.go#L99) since it doesn't have the [DotNL](https://golang.org/pkg/regexp/syntax/#DotNL) regexp flag enabled.

Reply:
Thanks for looking into this. Maybe @beorn7 and @peterbourgon have some ideas here.

Reply:
Yep, the stdlib adapter doesn't handle newlines. PR is welcome :)

Reply:
@SuperQ If I understand @peterbourgon correctly, he suggests a PR for the stdlib adapter, not for prometheus/client_golang.

See my comment there: https://github.com/prometheus/client_golang/pull/814#issuecomment-727599241

Reply:
My intention is that we do both. I don't like newlines in errors, as it complicates logging.

Reply:
I think we can go with an option to customize the formatting in prometheus/client_golang. In that way, we won't break anyone.

Reply:
Sure, we can chat about it sometime soon. This bug has been here for a while.

Reply:
Upstream go-kit PR: https://github.com/go-kit/kit/pull/1041

Reply:
After some thinking, I have to admit that we cannot add custom error formatting without a breaking change. However, it would be quite easy to create a custom `Logger` that intercepts the `MultiError` types and formats them into a single line. Happy to help if you want to do that in node_exporter.

See https://github.com/prometheus/client_golang/pull/821 for better documentation and https://github.com/prometheus/client_golang/pull/820 to change the behavior in v2.